File: packages/solid-fuse/src/focus-node.ts
import { createSignal } from "solid-js";
import { createHandle, type Handle } from "./handle";

export type FocusNode = Handle<"focusNode"> & {
  hasFocus: () => boolean;
  focus: () => void;
  unfocus: () => void;
  dispose: () => void;
};

export function createFocusNode(): FocusNode {
  const [hasFocus, setHasFocus] = createSignal(false);
  const { node, call, dispose } = createHandle("focusNode", { setHasFocus });
  return {
    node,
    hasFocus,
    focus: () => call("focus"),
    unfocus: () => call("unfocus"),
    dispose,
  };
}

File: packages/solid-fuse/dart/lib/src/handles/focus_node.dart
import 'package:flutter/material.dart';

import '../fuse_handle.dart';

class FuseFocusNode extends FuseHandle<FocusNode> {
  FuseFocusNode(super.node) : object = FocusNode() {
    object.addListener(
      () => node.function('setHasFocus')?.call(object.hasFocus),
    );
  }

  @override
  final FocusNode object;

  @override
  Future<dynamic> call(String method, dynamic value) async {
    switch (method) {
      case 'focus':
        object.requestFocus();
      case 'unfocus':
        object.unfocus();
      default:
        throw StateError('Unknown focusNode method: $method');
    }
    return null;
  }

  @override
  void dispose() => object.dispose();
}
